File: src/main/java/ru/dushkinmir/absolutelyRandom/sex/SexManager.java
package ru.dushkinmir.absolutelyRandom.sex;

import net.kyori.adventure.text.Component;
import org.bukkit.Bukkit;
import org.bukkit.Location;
import org.bukkit.entity.Player;
import org.bukkit.plugin.Plugin;
import org.bukkit.scoreboard.Scoreboard;
import org.bukkit.scoreboard.ScoreboardManager;
import org.bukkit.scoreboard.Team;
import org.bukkit.util.Vector;
import ru.dushkinmir.absolutelyRandom.utils.PlayerUtils;

import java.util.Random;

public class SexManager {

    private static final CollisionManager collisionManager = new CollisionManager();
    private static final long SEX_DURATION = 15;

    public static void triggerSexEvent(Player initiator, Player targetPlayer, Plugin plugin, AnalFissureHandler fissureHandler) {

        // Проверяем, что целевой игрок существует и в сети
        if (targetPlayer == null || !targetPlayer.isOnline()) {
            PlayerUtils.sendMessageToPlayer(
                    initiator,
                    Component.text("Игрок не найден или не в сети."),
                    PlayerUtils.MessageType.CHAT);
            return;
        }

        // Проверяем расстояние между игроками
        if (initiator.getLocation().distance(targetPlayer.getLocation()) > 1.5) {
            PlayerUtils.sendMessageToPlayer(
                    initiator,
                    Component.text("Игрок слишком далеко. Подойдите ближе."),
                    PlayerUtils.MessageType.CHAT);
            return;
        }

        // Отключаем коллизии для обоих игроков
        collisionManager.addPlayerToNoCollision(initiator);
        collisionManager.addPlayerToNoCollision(targetPlayer);

        // Определяем, кто из них будет двигаться
        boolean initiatorMoves = new Random().nextBoolean(); // true - движется инициатор, false - целевой игрок

        if (initiatorMoves) {
            performMovement(initiator, targetPlayer, plugin);
            fissureHandler.checkForFissure(targetPlayer); // Проверяем фиссуры для целевого игрока
        } else {
            performMovement(targetPlayer, initiator, plugin);
            fissureHandler.checkForFissure(initiator); // Проверяем фиссуры для инициатора
        }

        // Включаем коллизии обратно через 15 секунд
        Bukkit.getScheduler().runTaskLater(
                plugin,
                () -> {
                    collisionManager.removePlayerFromNoCollision(initiator);
                    collisionManager.removePlayerFromNoCollision(targetPlayer);
                },
                SEX_DURATION * 20L // 15 секунд
        );
    }

    private static void performMovement(Player movingPlayer, Player stationaryPlayer, Plugin plugin) {
        teleportBehind(stationaryPlayer, movingPlayer);
        movePlayer(movingPlayer, plugin);

        PlayerUtils.sendMessageToPlayer(
                movingPlayer,
                Component.text("<%s> даа мой папочка, накажи меня!!~".formatted(stationaryPlayer.getName())),
                PlayerUtils.MessageType.CHAT);
        PlayerUtils.sendMessageToPlayer(
                stationaryPlayer,
                Component.text("<%s> ода, давай сучка, нагибайся)".formatted(movingPlayer.getName())),
                PlayerUtils.MessageType.CHAT);
    }


    // Метод для движения игрока вперед-назад на ограниченное время (в секундах)
    private static void movePlayer(Player player, Plugin plugin) {
        player.sendMessage("Вы будете двигаться вперед-назад!");

        // Устанавливаем движение вперед-назад
        int taskId = Bukkit.getScheduler().runTaskTimerAsynchronously(
                plugin,
                new Runnable() {
                    private boolean moveForward = true;

                    @Override
                    public void run() {
                        if (!player.isOnline()) {
                            return;
                        }

                        // Получаем вектор скорости игрока (текущий вектор движения)
                        Location currentLocation = player.getLocation();
                        Vector direction = currentLocation.getDirection().normalize(); // Получаем направление взгляда и нормализуем

                        // Обнуляем вертикальную составляющую
                        direction.setY(0); // Убираем вертикальную составляющую
                        direction = direction.normalize(); // Нормализуем снова после обнуления Y


                        if (moveForward) {
                            player.setVelocity(direction.multiply(0.25)); // Двигаем вперед
                        } else {
                            player.setVelocity(direction.multiply(-0.25)); // Двигаем назад
                        }

                        moveForward = !moveForward; // Меняем направление
                    }
                },
                0L, // Начало сразу
                5L  // Выполняем каждые 5 тиков (1/4 секунды)
        ).getTaskId();

        // Останавливаем задачу через 15 секунд (300 тиков)
        Bukkit.getScheduler().runTaskLater(
                plugin,
                () -> Bukkit.getScheduler().cancelTask(taskId),
                SEX_DURATION * 20L // 15 секунд = 300 тиков
        );
    }    // Метод для телепортации игрока "движущегося" за спину "стоячего"

    private static void teleportBehind(Player stationaryPlayer, Player movingPlayer) {
        Location stationaryLocation = stationaryPlayer.getLocation();
        Vector backwardDirection = stationaryLocation.getDirection().multiply(-1); // Вектор, указывающий назад

        // Рассчитываем новую позицию для телепортации игрока на 1.5 блока позади
        Location teleportLocation = stationaryLocation.clone().add(backwardDirection.normalize().multiply(1.1));
        teleportLocation.setY(stationaryLocation.getY()); // Оставляем высоту на том же уровне

        // Телепортируем игрока
        movingPlayer.teleport(teleportLocation);
    }

    private static class CollisionManager {

        private Team noCollisionTeam;

        public CollisionManager() {
            setupNoCollisionTeam();
        }

        // Создаём команду с отключёнными коллизиями
        private void setupNoCollisionTeam() {
            ScoreboardManager manager = Bukkit.getScoreboardManager();
            Scoreboard board = manager.getMainScoreboard();

            // Проверяем, существует ли команда уже
            noCollisionTeam = board.getTeam("noCollision");
            if (noCollisionTeam == null) {
                // Создаём новую команду с именем "noCollision"
                noCollisionTeam = board.registerNewTeam("noCollision");
                noCollisionTeam.setOption(Team.Option.COLLISION_RULE, Team.OptionStatus.NEVER); // Отключаем коллизии
            }
        }

        // Добавляем игрока в команду без коллизий
        public void addPlayerToNoCollision(Player player) {
            if (noCollisionTeam != null && !noCollisionTeam.hasEntry(player.getName())) {
                noCollisionTeam.addEntry(player.getName());
            }
        }

        // Удаляем игрока из команды (включение коллизий)
        public void removePlayerFromNoCollision(Player player) {
            if (noCollisionTeam != null && noCollisionTeam.hasEntry(player.getName())) {
                noCollisionTeam.removeEntry(player.getName());
            }
        }
    }
}

File: src/main/java/ru/dushkinmir/absolutelyRandom/features/sex/SexManager.kt
package ru.dushkinmir.absolutelyRandom.features.sex

import net.kyori.adventure.text.Component
import net.kyori.adventure.text.format.NamedTextColor
import net.kyori.adventure.text.format.TextDecoration
import org.bukkit.Bukkit
import org.bukkit.entity.Player
import org.bukkit.plugin.Plugin
import org.bukkit.scheduler.BukkitRunnable
import org.bukkit.scoreboard.Team
import ru.dushkinmir.absolutelyRandom.features.actions.actions.Group.FallingBlocksTask
import ru.dushkinmir.absolutelyRandom.utils.PlayerUtils
import ru.dushkinmir.absolutelyRandom.utils.ui.ChatConfirmation
import java.util.*

object SexManager {

    private val collisionManager = CollisionManager()
    private const val SEX_DURATION: Long = 15

    fun triggerSexEvent(initiator: Player, targetPlayer: Player, plugin: Plugin, fissureHandler: AnalFissureHandler) {
        // Проверяем, что целевой игрок существует и в сети
        if (!targetPlayer.isOnline) {
            PlayerUtils.sendMessageToPlayer(
                initiator,
                Component.text("Игрок не найден или не в сети."),
                PlayerUtils.MessageType.CHAT
            )
            return
        }
        val chatConfirmation = ChatConfirmation(plugin)

        chatConfirmation.showConfirmation(
            targetPlayer,
            Component.text("${initiator.name} хочет с тобой почпокаться! примешь ли ты его?")
                .color(NamedTextColor.YELLOW),
            onConfirm = {
                PlayerUtils.sendMessageToPlayer(
                    initiator,
                    Component.text("ура! ${targetPlayer.name} принял твое предложение!")
                        .color(NamedTextColor.GREEN) // Позитивное подтверждение, зеленый
                        .decorate(TextDecoration.BOLD), // Жирный текст,
                    PlayerUtils.MessageType.CHAT
                )
                PlayerUtils.sendMessageToPlayer(
                    targetPlayer,
                    Component.text("ты принял предложение, щас будет трахен трахен чпокен чпокен")
                        .color(NamedTextColor.AQUA) // Синие, немного игривые
                        .decorate(TextDecoration.ITALIC), // Курсив,
                    PlayerUtils.MessageType.CHAT
                )
                object : BukkitRunnable() {
                    override fun run() {
                        executeSex(initiator, targetPlayer, plugin, fissureHandler)
                    }
                }.runTaskLater(plugin, 20L)
            },
            onCancel = {
                PlayerUtils.sendMessageToPlayer(
                    initiator,
                    Component.text("увы, но ${targetPlayer.name} отверг твое предложение(((\nлох")
                        .color(NamedTextColor.RED), // Отрицательное сообщение, серый
                    PlayerUtils.MessageType.CHAT
                )
                PlayerUtils.sendMessageToPlayer(
                    targetPlayer,
                    Component.text("ты отказал ${initiator.name} в чпок-чпок!\nтуда его")
                        .color(NamedTextColor.GRAY), // Отказ, красный,
                    PlayerUtils.MessageType.CHAT
                )
            }
        )
    }

    private fun executeSex(
        initiator: Player,
        targetPlayer: Player,
        plugin: Plugin,
        fissureHandler: AnalFissureHandler
    ) {

        // Проверяем расстояние между игроками
        if (initiator.location.distance(targetPlayer.location) > 1.5) {
            PlayerUtils.sendMessageToPlayer(
                initiator,
                Component.text("Игрок слишком далеко. Подойдите ближе."),
                PlayerUtils.MessageType.CHAT
            )
            return
        }

        // Отключаем коллизии для обоих игроков
        collisionManager.addPlayerToNoCollision(initiator)
        collisionManager.addPlayerToNoCollision(targetPlayer)

        // Определяем, кто из них будет двигаться
        val initiatorMoves = Random().nextBoolean() // true - движется инициатор, false - целевой игрок

        if (initiatorMoves) {
            performMovement(initiator, targetPlayer, plugin, fissureHandler)
        } else {
            performMovement(targetPlayer, initiator, plugin, fissureHandler)
        }

        // Включаем коллизии обратно через 15 секунд
        object : BukkitRunnable() {
            override fun run() {
                collisionManager.removePlayerFromNoCollision(initiator)
                collisionManager.removePlayerFromNoCollision(targetPlayer)
            }
        }.runTaskLater(plugin, SEX_DURATION * 20L)
    }

    private fun performMovement(
        movingPlayer: Player,
        stationaryPlayer: Player,
        plugin: Plugin,
        fissureHandler: AnalFissureHandler
    ) {
        teleportBehind(stationaryPlayer, movingPlayer)
        movePlayer(movingPlayer, stationaryPlayer, plugin, fissureHandler)

        PlayerUtils.sendMessageToPlayer(
            movingPlayer,
            Component.text("<${stationaryPlayer.name}> даа мой папочка, накажи меня!!~"),
            PlayerUtils.MessageType.CHAT
        )
        PlayerUtils.sendMessageToPlayer(
            stationaryPlayer,
            Component.text("<${movingPlayer.name}> ода, давай сучка, нагибайся)"),
            PlayerUtils.MessageType.CHAT
        )
    }


    // Метод для движения игрока вперед-назад на ограниченное время (в секундах)
    private fun movePlayer(
        movingPlayer: Player,
        stationaryPlayer: Player,
        plugin: Plugin,
        fissureHandler: AnalFissureHandler
    ) {
        movingPlayer.sendMessage("Вы будете двигаться вперед-назад!")

        // Устанавливаем движение вперед-назад
        val taskId = Bukkit.getScheduler().runTaskTimerAsynchronously(
            plugin,
            object : Runnable {
                private var moveForward = true

                override fun run() {
                    if (!movingPlayer.isOnline) {
                        return
                    }

                    // Получаем вектор скорости игрока (текущий вектор движения)
                    val currentLocation = movingPlayer.location
                    var direction = currentLocation.direction.normalize() // Получаем направление взгляда и нормализуем

                    // Обнуляем вертикальную составляющую
                    direction.y = 0.0 // Убираем вертикальную составляющую
                    direction = direction.normalize() // Нормализуем снова после обнуления Y


                    if (moveForward) {
                        movingPlayer.velocity = direction.multiply(0.35) // Двигаем вперед
                    } else {
                        movingPlayer.velocity = direction.multiply(-0.35) // Двигаем назад
                    }

                    moveForward = !moveForward // Меняем направление
                }
            },
            0L, // Начало сразу
            2L
        ).taskId

        // Останавливаем задачу через 15 секунд (300 тиков)
        object : BukkitRunnable() {
            override fun run() {
                Bukkit.getScheduler().cancelTask(taskId)
                fissureHandler.checkForFissure(stationaryPlayer)
                FallingBlocksTask(plugin, ArrayList(listOf(movingPlayer, stationaryPlayer))).runTaskTimer(
                    plugin,
                    0L,
                    1L
                )
            }
        }.runTaskLater(plugin, SEX_DURATION * 20L)
    }

    // Метод для телепортации игрока "движущегося" за спину "стоячего"
    private fun teleportBehind(stationaryPlayer: Player, movingPlayer: Player) {
        val stationaryLocation = stationaryPlayer.location
        val backwardDirection = stationaryLocation.direction.multiply(-1.0) // Вектор, указывающий назад

        // Рассчитываем новую позицию для телепортации игрока на 1.5 блока позади
        val teleportLocation = stationaryLocation.clone().add(backwardDirection.normalize().multiply(1.1))
        teleportLocation.y = stationaryLocation.y // Оставляем высоту на том же уровне

        // Телепортируем игрока
        movingPlayer.teleport(teleportLocation)
    }

}

private class CollisionManager {

    private var noCollisionTeam: Team?

    init {
        noCollisionTeam = setupNoCollisionTeam()
    }

    // Создаём команду с отключёнными коллизиями
    private fun setupNoCollisionTeam(): Team? {
        val manager = Bukkit.getScoreboardManager()
        val board = manager.mainScoreboard

        // Проверяем, существует ли команда уже
        var team = board.getTeam("noCollision")
        if (team == null) {
            // Создаём новую команду с именем "noCollision"
            team = board.registerNewTeam("noCollision")
            team.setOption(Team.Option.COLLISION_RULE, Team.OptionStatus.NEVER) // Отключаем коллизии
        }
        return team
    }

    // Добавляем игрока в команду без коллизий
    fun addPlayerToNoCollision(player: Player) {
        if (noCollisionTeam != null && !noCollisionTeam!!.hasEntry(player.name)) {
            noCollisionTeam!!.addEntry(player.name)
        }
    }

    // Удаляем игрока из команды (включение коллизий)
    fun removePlayerFromNoCollision(player: Player) {
        if (noCollisionTeam != null && noCollisionTeam!!.hasEntry(player.name)) {
            noCollisionTeam!!.removeEntry(player.name)
        }
    }
}
